##                                       Plantix Case Study

# KPI's that one should use to measure the performance of app:-
## General Mobile app KPI
### 1.User Growth Rate
### 2.Mobile Downloads
### 3.Registrations
### 4. Subscriptions
### 5.Installs/Unistalls

## App Engagement Metrices
### 1.Retention Rate
### 2.Daily Active Users
### 3.Monthly Active Users
### 4.Social Shares
### 5.Brand Awareness

## Revenue Focused KPI
### 1.Avg revenue per user
### 2.Purchases
### 3.Customer Acquistion Cost
### 4.Return on investment
### 5.Cost per install

## User experience KPI
### 1.Load Time
### 2.Devices
### 3.Carriers
### 4.Permissions Granted
### 5.API Latency

## App Atore OPtimization
### 1.Keywords
### 2.App Store Category Ranking
### 3.Views To Install
### 4. Ratings
### 5.Reviews


# Importing Important Libraries

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Reading all the files 

In [2]:
login_data=pd.read_csv("login_logs.csv")
login_data.head()

login_log_id  user_id             login_time
0       1385328    99160  2021-07-01 00:00:33.0
1       1385329    83272  2021-07-01 00:00:38.0
2       1385330    83272  2021-07-01 00:00:42.0
3       1385331    96640  2021-07-01 00:02:08.0
4       1385332    99160  2021-07-01 00:02:38.0

In [3]:
login_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 666357 entries, 0 to 666356
Data columns (total 3 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   login_log_id  666357 non-null  int64 
 1   user_id       666357 non-null  int64 
 2   login_time    666357 non-null  object
dtypes: int64(2), object(1)
memory usage: 15.3+ MB


In [4]:
sales=pd.read_csv("sales_orders.csv")
sales.head()

order_id  fk_buyer_id  fk_depot_id sales_order_status  \
0    119874        82228            5           Rejected   
1    119877        40476            1           Rejected   
2    119880        41728            8            Shipped   
3    119883        97616            4           Rejected   
4    119886        12440            1           Rejected   

           creation_time  
0  2021-07-01 00:35:28.0  
1  2021-07-01 02:02:43.0  
2  2021-07-01 02:42:12.0  
3  2021-07-01 02:56:00.0  
4  2021-07-01 03:11:31.0

In [5]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13630 entries, 0 to 13629
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            13630 non-null  int64 
 1   fk_buyer_id         13630 non-null  int64 
 2   fk_depot_id         13630 non-null  int64 
 3   sales_order_status  13630 non-null  object
 4   creation_time       13630 non-null  object
dtypes: int64(3), object(2)
memory usage: 532.5+ KB


In [6]:
items_data=pd.read_csv("sales_orders_items.csv")
items_data.head()


order_item_id  fk_product_id  fk_order_id  ordered_quantity  \
0          62609           7818       119874                 1   
1          62610           7607       119877                 1   
2          62611           9916       119880                 2   
3          62612          11118       119880                 2   
4          62613          12702       119880                 2   

   order_quantity_accepted     rate  
0                        0  4993.00  
1                        0  6203.40  
2                        2  2904.41  
3                        2  2205.00  
4                        2  2200.85

In [7]:
items_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20488 entries, 0 to 20487
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   order_item_id            20488 non-null  int64  
 1   fk_product_id            20488 non-null  int64  
 2   fk_order_id              20488 non-null  int64  
 3   ordered_quantity         20488 non-null  int64  
 4   order_quantity_accepted  20488 non-null  int64  
 5   rate                     20488 non-null  float64
dtypes: float64(1), int64(5)
memory usage: 960.5 KB


### Extracting year creation_year

In [8]:
sales['creationyear'] = pd.DatetimeIndex(sales['creation_time']).year
sales['creationyear']
del sales['creation_time']

In [9]:
sales.head()

order_id  fk_buyer_id  fk_depot_id sales_order_status  creationyear
0    119874        82228            5           Rejected          2021
1    119877        40476            1           Rejected          2021
2    119880        41728            8            Shipped          2021
3    119883        97616            4           Rejected          2021
4    119886        12440            1           Rejected          2021

### Finding shipments in year 2021  and year 2022

In [10]:

df2=sales.query("sales_order_status == 'Shipped'")
df3=df2.query("creationyear == 2021")
ship2021=df3.shape[0]
print(f"items shipped in 2021 :",ship2021)
ship2022=len(df2)-ship2021
print(f"items shipped in 2022 :",ship2022)


items shipped in 2021 : 2434
items shipped in 2022 : 3489


### answer (2.a) As the products shipped in 2022 has been increased this shows business growth.

### Extracting year from login_time column

In [11]:
login_data['year'] = pd.DatetimeIndex(login_data['login_time']).year
login_data['year']
del login_data['login_time']


In [12]:
login_data.tail()

login_log_id  user_id  year
666352       4670509   207632  2022
666353       4670510   207632  2022
666354       4670511   207632  2022
666355       4670512   111788  2022
666356       4670513   111788  2022

### Finding logins in year 2021

In [13]:
count=0
for i in login_data['year']:
    if(i==2021):
        count=count+1
print(count)

271240


### Finding logins in year 2022

In [14]:
login2021=271240
login2022=(len(login_data["year"])-login2021)
growth_login=login2022-login2021
print(growth_login)

123877


# answer(2.c) As we can clearly see that the logins  has increased by 123877 from year 2021 to 2022 so this shows that our user base grows

### Inserting  year column to items data  

In [15]:
extracted_col=sales['creationyear']
print(extracted_col)
items_data.insert(0,'colnew',extracted_col)


0        2021
1        2021
2        2021
3        2021
4        2021
         ... 
13625    2022
13626    2022
13627    2022
13628    2022
13629    2022
Name: creationyear, Length: 13630, dtype: int64


In [16]:
items_data.head()

colnew  order_item_id  fk_product_id  fk_order_id  ordered_quantity  \
0  2021.0          62609           7818       119874                 1   
1  2021.0          62610           7607       119877                 1   
2  2021.0          62611           9916       119880                 2   
3  2021.0          62612          11118       119880                 2   
4  2021.0          62613          12702       119880                 2   

   order_quantity_accepted     rate  
0                        0  4993.00  
1                        0  6203.40  
2                        2  2904.41  
3                        2  2205.00  
4                        2  2200.85

## Filling Null Values using mode

In [26]:
items_data['colnew'] = items_data['colnew'].fillna(items_data['colnew'].mode()[0])

In [27]:
items_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20488 entries, 0 to 20487
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   colnew                   20488 non-null  float64
 1   order_item_id            20488 non-null  int64  
 2   fk_product_id            20488 non-null  int64  
 3   fk_order_id              20488 non-null  int64  
 4   ordered_quantity         20488 non-null  int64  
 5   order_quantity_accepted  20488 non-null  int64  
 6   rate                     20488 non-null  float64
dtypes: float64(2), int64(5)
memory usage: 1.1 MB


### Finding top products details of Year 2021 and Year 2022

In [28]:
topsales1=items_data.query("colnew == 2021")
topsales1.shape
top5=topsales1.sort_values(by=['ordered_quantity'], ascending=False)
print("TOP 5 products details of 2021")
top5.head()


TOP 5 products details of 2021


colnew  order_item_id  fk_product_id  fk_order_id  ordered_quantity  \
2842  2021.0          65451           7666       125169               999   
5017  2021.0          67626           7533       129360               481   
1237  2021.0          63846           1040       122208               360   
5176  2021.0          67785          12800       129666               300   
5886  2021.0          68495           2862       131055               200   

      order_quantity_accepted      rate  
2842                        0  10182.00  
5017                        0   4882.00  
1237                        0   3570.85  
5176                        0      4.00  
5886                        0   4200.00

In [29]:
topsales2=items_data.query("colnew == 2022")
topsales2.shape
top=topsales2.sort_values(by=['ordered_quantity'], ascending=False)
print("TOP 5 products details of 2022")
top.head()

TOP 5 products details of 2022


colnew  order_item_id  fk_product_id  fk_order_id  ordered_quantity  \
11624  2022.0         148725           1043       300192               835   
19792  2022.0         156893          12547       317475               500   
9191   2022.0          71800           8210       137280               450   
6732   2022.0          69341           8219       132573               400   
10197  2022.0         147298           1041       297468               366   

       order_quantity_accepted     rate  
11624                        0  22849.0  
19792                      500   5730.0  
9191                         0  12674.0  
6732                         0  13288.2  
10197                        0   4954.0

### answer(4)  The first Problem in july 2021 data is the orders have been rejected more than the orders that are shipped.
#### To conquer this problem we need to first understand the behaviour     of the rejected orders and reasons for rejections.
#### 1. To add reasons column when someone is trying to cancel order.
#### 2. To add expectation column to get what customer actually wants.
#### 3. A good recommendation system to recommend similar and options that satisfies customer needs.
#### 4. To maintain data for customers who are canceling order to understand that the customer is genuine or not.

### The second problem is inability to meet customer demands in terms of quantity.
#### 1.Inventory is an essential aspect whether we are manufacturing units or suppliers.
#### 2.The suggestion is to build multi layer monitoring over the inventories so that in case of mistake or failure from one layer there is still acurate result about quantity available.
#### 3.The multil  layer monitoring system should be less manual and includes the advancement of technology like automatic updates in data whenever there are changes.
#### 4.An appropiate dashboard should be prepared which focus on the products which are highly popular among customers. We should never lack in that. Visualisation tools such as Power Bi or excel dashboards can be one of the approach.
#### 5. Information regarding quantity available of the product he/she viewing at  should be visible so that the customer knows before hand. This can save us from the negative image customer built in their mind.



### Finding correlation between login frequency and order quantity

In [19]:
login_data['login_log_id'].corr(items_data['ordered_quantity'])

-0.009222960583501532

### answer(5) No the login frequencies doesn't really affect the order quantity